U4-9792 -- Update Umbraco to be FIPS compliant

protherj opened the pull request:
Set of commits to refactor code/API to allow for FIPS compliance.  

**Umbraco.Core/StringExtensions.cs**

- Create a new GenerateHash() String extensions method to detect if FIPS is required and use SHA1 for hashing if it is required
- Update the ToGuid() method to use the new GenerateHash() method.

**Umbraco.Web/UmbracoHelper.cs**

- Add a new CreateHash() method that uses the GenerateHash() String extension method.  
- Mark the CreateMd5Hash() method as obsolete and hidden from editor intellisense.

**umbraco.presentation/library.cs**

- Add a new CreateHash() method that uses the GenerateHash() String extension method.  
- Mark the md5() method as obsolete and hidden from editor intellisense.

**umbraco.MacroEngines/RazorCore/RazorMacroEngine.cs**

- Add a new GetHash() method that uses the GenerateHash() String extension method.  
- Mark the GetMd5() method as obsolete and hidden from editor intellisense.
- Update calls to GetMd5() to GetHash()

**Umbraco.Core/Security/BackOfficeUserStore.cs**

- Change call from ToMd5() to GenerateHash()

**Umbraco.Web/Models/Mapping/UserModelMapper.cs**

- Change call from ToMd5() to GenerateHash()

**Umbraco.Web/umbraco.presentation/umbraco/developer/Packages/installer.aspx.cs**

- Change call from library.md5() to library.CreateHash()

**Umbraco.Web/umbraco.presentation/umbraco/plugins/tinymce3/GzipCompressor.cs**

- Removed Md5() method
- Add a new Hash() method that uses the GenerateHash() String extension method.  
- Change call from Md5() to Hash()

**UmbracoExamine/LocalStorage/AzureLocalStorageDirectory.cs**

- Change call from ToMd5() to GenerateHash()

**Umbraco.Web/Editors/GravatarController.cs**

- Check if FIPS is required.  If it is, then return null and don't retrieve the Gravatar image.

**umbraco.editorControls/tinyMCE3/webcontrol/plugin/GzipCompressor.cs**

- Change call from ToMd5() to GenerateHash()

**Examine**

- Updated the Examine Nuget to the current latest stable version, 0.1.83, that adds FIPS detection and compliant hashing algorithms.

Shazwazza commented on src/Umbraco.Web/Editors/GravatarController.cs:
```diff
@@ -13,6 +14,14 @@ public class GravatarController : UmbracoAuthorizedJsonController
+            // and the website will not run.
+            if (CryptoConfig.AllowOnlyFipsAlgorithms)
+            {
+                return null;
```
what happens in the UI when this is null, is there JS errors in the console?

protherj replied:
No JS errors present and the request returns a status of 200.  The image simply stays a teal "U" Umbraco logo.

protherj commented:
Updated the Examine Nuget to the 0.1.83 version.

protherj commented:
Updated Unit Tests to test against of 0.1.83 Examine Nuget version

protherj commented:
Latest testing status here: http://issues.umbraco.org/issue/U4-9792

protherj commented on src/Umbraco.Core/StringExtensions.cs:
```diff
@@ -702,65 +702,70 @@ public static object ParseInto(this string val, Type type)
+        private static string GenerateHash(this string str, string hashType)
+        {
+            //create an instance of the correct hashing provider based on the type passed in
+            var hasher = HashAlgorithm.Create(hashType);
```
I see what you did for the Examine project and will copy that approach.

protherj replied:
So I have the hasher in a using after the null check and that will call finally on that instance.  So that should call dispose and clean up I believe.  How else could it check for null then be in a using?

Thanks,
Jason

protherj replied:
Just a note: we talked about this at Codegarden '17 and my understanding is that this is good to go.  However, let me know if there is anything further I need to do on this.

Thanks,
Jason

protherj commented on src/Umbraco.Core/StringExtensions.cs:
```diff
@@ -1465,10 +1470,7 @@ internal static string ToValidXmlString(this string text)
-            byte[] myStringBytes = Encoding.ASCII.GetBytes(text);
-            byte[] hash = md5.ComputeHash(myStringBytes);
-            return new Guid(hash);
+            return new Guid(text.GenerateHash());
```
According to this SHA1 should work, but I will need to edit this to truncate the hash to 128 bits

https://stackoverflow.com/a/6248764/32908

protherj replied:
What about adding something like this?

 https://github.com/LogosBible/Logos.Utility/blob/master/src/Logos.Utility/GuidUtility.cs#L34

See discussion here:

https://stackoverflow.com/a/5657517/32908

protherj replied:
OK, I added this method and forced SHA-1 and the unit tests passed and everything looks like a valid GUID.  In fact, it may be more valid because it actually adheres to the actually RFC 4122 spec.

I'm pushing this to Github now and will update this pull request.

protherj commented:
OK, I've updated the PR.  I believe I've addressed everything that was pending (the ToGuid() issue).  

Please let me know if I need to update anything and I will turn around the change as fast as possible.

Thanks,
Jason
